### PR TITLE
Single-group Interrupted Time-Series Analysis (ITSA)

### PR DESCRIPTION
**Pull request by alashi-ms**
# Summary
This branch introduces new function `create_ITSA()` which estimates the intervention's effects (e.g. reOrg, shift to remote work, policy change, training etc.) on WpA metrics.

# Changes
The changes made in this PR are:
   1. Added `create_ITSA()` as a function for estimating effects of the intervention on WpA metrics. The necessary requirement is 
       to know the date or date range of the intervention.  

# Checks
- [ ] All R CMD checks pass 
- [ ] `roxygen2::roxygenise()` has been run prior to merging to ensure that `.Rd` and `NAMESPACE` files are up to date.
- [ ] Update `NEWS.md` 

# Notes

### CODE REVIEW

**Review by martinctc — COMMENTED** [review]
Thanks @alashi-ms! This is looking amazing. I have some comments for now, please let me know if they make sense / or are something you'd like to implement. I'd be happy to take some of these:
- `YYYY-MM-DD` format in the documentation seems to be inconsistent with what is being used in the function
- Would prefer if we can reduce the package dependencies of `lmtest`, `portes`, and `sandwich`. I can attempt this if it is not easy fix
- Can make the ITSA plots more informative with plot titles, captions, etc. 
- I would add yourself as the author of this function. You can reference <https://github.com/microsoft/wpa/blob/main/R/network_describe.R>

Thank you so much!

File: R/create_ITSA.R [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+          y = Y_at_intervantion_when_intervantion_happened + dY,
+          label = c(paste0("beta_2=", round(beta_2, 3), collapse=" "))
+        )
+
+        # Create final plot
+        p_final <- p +
+          geom_vline(xintercept=event_T, color="red", size=1) +
+          annotate("segment", x = pos_x_start_beta_2, xend = pos_x_end_beta_2, y = pos_y_start_beta_2, yend = pos_y_end_beta_2, colour = "black", size=0.5, alpha=0.6, arrow=arrow())+
+          annotate("segment", x = pos_x_end_beta_2, xend = pos_x_end_beta_2, y = Y_at_intervantion_when_no_intervantion_happened, yend = Y_at_intervantion_when_intervantion_happened, colour = "purple", size=2, alpha=1)+
+          geom_text(data=annotation, aes( x=x, y=y, label=label), color="orange", size=5, angle=0, fontface="bold" )
```
[the remark this answers is not in the document]
Comment (alashi-ms) [reply R/create_ITSA.R:308]: let's commit to it after we get users' feedbacks

File: DESCRIPTION [hunk 2]
```diff
@@ -54,7 +54,10 @@ Imports:
     proxy,
     ggwordcloud,
     methods,
-    data.table
+    data.table,
+    lmtest,
+    portes,
+    sandwich
```
[the remark this answers is not in the document]
Comment (alashi-ms) [reply DESCRIPTION:60]: At the moment we need to keep these packages as they do heavy math and reimplementation of their functionality would be time consuming.

[the remark this answers is not in the document]
Comment (martinctc) [reply DESCRIPTION:60]: This is fine @alashi-ms - I will keep them as optional dependencies, i.e. users only need to install them if they use this specific function. Will work on this

**Review by m-m-powers — COMMENTED** [review]
Aleksey, this seems like a very powerful analysis! I made some general comments on things that didn't work for me or seemed inconsistent.

One thing to consider is that this function's output is difficult to decipher. It presupposes knowledge of that ITSA method you cite in the header. Is there a way to make the table return or the plot returns more interpretable by someone who isn't as familiar? For example, you say in the instructions to look for p-values that are significant in beta_2 or beta_3. Could you instead do that check for the analyst? If the analyst input their significance level, couldn't the function just return the significant metrics? Or at least differentiate between those that are significant and those that aren't?

File: R/create_ITSA.R [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+           before_start = min(as.Date(data$Date, "%m/%d/%Y")),
+           before_end,
+           after_start,
+           after_end = max(as.Date(data$Date, "%m/%d/%Y")),
+           ac_lags_max = 7,
+           return = 'table') {
+
+    ## Check inputs types
+    assert_that(is.data.frame(data) &&
+                is.character(before_start) &&
```
[the remark this answers is not in the document]
Comment (alashi-ms) [reply R/create_ITSA.R]: fixed by adding extra check for Date type is.character(before_start)|inherits(before_start, "Date")

File: R/create_ITSA.R [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+          geom_point(aes(y = Y), color = "blue")+
+          geom_line(data=hat_Y_before_and_at_intervention, aes(x=T, y=Y), size=1)+
+          geom_line(data=hat_Y_after_and_at_intervention, aes(x=T, y=Y), size=1)
+
+        # Calculate plotting area range and scale
+        dY <- (max(data_OLS[,"Y"]) - min(data_OLS[,"Y"]))/10
+        dX <- 1
+
+        # Calculate annotation positions on the graph
+        Y_at_intervantion_when_no_intervantion_happened <- hat_Y_before_and_at_intervention[dim(hat_Y_before_and_at_intervention)[1] ,"Y"]
```
[the remark this answers is not in the document]
Comment (alashi-ms) [reply R/create_ITSA.R]: fixed spelling in https://github.com/microsoft/wpa/pull/116/commits/b310a1ece635800202f1445c0a5954a8551d1820

File: R/create_ITSA.R [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+                          error_warning = paste0('Error: ',c$message, "; lm_train_success=", lm_train_success_flag, collapse = " "))
+      },
+      warning = function(c){
+        buf <- data.frame(metric_name = metric_name,
+                          beta_2 = ifelse(exists("beta_2", inherits = FALSE), beta_2, -1),
+                          beta_3 = ifelse(exists("beta_3", inherits = FALSE), beta_3, -1),
+                          beta_2_pvalue = ifelse(exists("beta_2_pvalue", inherits = FALSE), beta_2_pvalue, -1),
+                          beta_3_pvalue = ifelse(exists("beta_3_pvalue", inherits = FALSE), beta_3_pvalue, -1),
+                          AR_flag = AR_flag,
+                          error_warning = paste0('Warning: ',c$message, "; lm_train_success=", lm_train_success_flag, collapse = " "))
```
[the remark this answers is not in the document]
Comment (alashi-ms) [reply R/create_ITSA.R:259]: Error message indicates that library portes (portes::LjungBox) is not loaded for some reason
53-Layer and 54 - HourlyRate metrics stays constant over time, this is why Warning message is generated.

File: R/create_ITSA.R [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+#' @description
+#' This function implements ITSA method described in the paper 'Conducting interrupted time-series analysis for
+#' single- and multiple-group comparisons', Ariel Linden, The Stata Journal (2015), 15, Number 2, pp. 480-500
+#'
+#' @param data Person Query as a dataframe including date column named 'Date'
+#' This function assumes the data format is MM/DD/YYYY as is standard in a WpA query output.
+#' @param before_start Start date of 'before' time period in YYYY-MM-DD format as character type. Defaults to earliest date in dataset.
+#' @param before_end End date of 'before' time period in YYYY-MM-DD  format as character type.
+#' @param after_start Start date of 'after' time period in YYYY-MM-DD  format as character type. Defaults to day after before_end.
+#' @param after_end End date of 'after' time period in YYYY-MM-DD  format as character type. Defaults to latest date in dataset.
```
[the remark this answers is not in the document]
Comment (alashi-ms) [reply R/create_ITSA.R]: this has been addressed in https://github.com/microsoft/wpa/pull/116/commits/5836f040739ccae0ff35e4eeca82a7132a59f5ef

File: R/create_ITSA.R [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+          y = Y_at_intervantion_when_intervantion_happened + dY,
+          label = c(paste0("beta_2=", round(beta_2, 3), collapse=" "))
+        )
+
+        # Create final plot
+        p_final <- p +
+          geom_vline(xintercept=event_T, color="red", size=1) +
+          annotate("segment", x = pos_x_start_beta_2, xend = pos_x_end_beta_2, y = pos_y_start_beta_2, yend = pos_y_end_beta_2, colour = "black", size=0.5, alpha=0.6, arrow=arrow())+
+          annotate("segment", x = pos_x_end_beta_2, xend = pos_x_end_beta_2, y = Y_at_intervantion_when_no_intervantion_happened, yend = Y_at_intervantion_when_intervantion_happened, colour = "purple", size=2, alpha=1)+
+          geom_text(data=annotation, aes( x=x, y=y, label=label), color="orange", size=5, angle=0, fontface="bold" )
```
[the remark this answers is not in the document]
Comment (alashi-ms) [reply R/create_ITSA.R:310]: added title to each plot https://github.com/microsoft/wpa/pull/116/commits/64da7f32a5247b81fb332837042087003352cef8

**Comment by alashi-ms** [comment]
> Aleksey, this seems like a very powerful analysis! I made some general comments on things that didn't work for me or seemed inconsistent.
> 
> One thing to consider is that this function's output is difficult to decipher. It presupposes knowledge of that ITSA method you cite in the header. Is there a way to make the table return or the plot returns more interpretable by someone who isn't as familiar? For example, you say in the instructions to look for p-values that are significant in beta_2 or beta_3. Could you instead do that check for the analyst? If the analyst input their significance level, couldn't the function just return the significant metrics? Or at least differentiate between those that are significant and those that aren't?

>>Could you instead do that check for the analyst? if the analyst input their significance level, couldn't the function just return the significant metrics? 
It is more convenient to see all results and then decide on the p-value threshold than specify it upfront. For example: If you set p-value threshold to too conservative level (e.g. 0.05 or smaller) you may end up with empty results and would need to re-run the script. On the other hand, filtering results in a spreadsheet saves time (no need for re-run) and provides full visibility on p-values making decision on threshold easier.  
>>Or at least differentiate between those that are significant and those that aren't? 
Significance depends on a risk an analyst in charge is willing to take and it varies from customer/scenario to customer/scenario. 
On the general note for the first iteration, I would prefer to follow low code concept without introducing additional parameters (e.g. beta_2_p_value_threshold, beta_3_p_value_threshold and some condition_flag indicating what combinations of the thresholds should be satisfied) and return to this discussion after collecting users feedback.

File: R/create_ITSA.R (hunk 6 above)
[the remark this answers is not in the document]
Comment (alashi-ms) [reply R/create_ITSA.R]: added the context https://github.com/microsoft/wpa/pull/116/commits/28e6bdb1c1e3b4adb7070b78adf679bc63094592

File: R/create_ITSA.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (m-m-powers) [reply R/create_ITSA.R:265]: I installed portes so not sure why it wouldn't show up. That's weird,probably my version of R

**Pull request merged**